### PR TITLE
feat: initial support for `npm@7`

### PR DESCRIPTION
**Pull request by G-Rath**
Sadly the audit output from `npm@7` is greatly reduced, meaning we don't have enough information to form a nice complete path to the vulnerable dependency on the tree.

I've implemented a "normalizer" that builds a `Finding` based on an advisory, from all auditing tools. This ensures standardization across all the tools, which'll be helpful if they switch to using the new `bulk` audit endpoint.

### CODE REVIEW

**Comment by G-Rath** [comment]
(I'll probably do a follow-up at some point refactoring most of the types so that the names are better aligned)

File: src/audit.ts [hunk 1]
```diff
@@ -101,6 +137,18 @@ const collectYarnAuditResults: AuditResultsCollector = async stdout => {
   });
 };
 
+const toMapOfFindings = (findings: Finding[]): Record<string, Finding> => {
```
Comment (fractalcactus) [inline src/audit.ts:140]: 🐘 in `Record<string` and `1Omit<Npm7Vulnerability`, does the `<` mean anything special, or is is just a character in the variable name? I'm just curious

File: src/types.ts [hunk 2]
```diff
@@ -3,7 +3,7 @@ export type Severity = 'info' | 'low' | 'moderate' | 'high' | 'critical';
 export type SeverityCounts = Record<Severity, number>;
 export type SeverityCountsWithTotal = Record<Severity | 'total', number>;
 
-export type Advisories = Record<string, Advisory>;
+export type Advisories = Record<string, Npm6Advisory>;
```
Comment (fractalcactus) [inline src/types.ts:6]: 🐘 
this is probably my lack of Js knowledge, what is going on with the code using `< >`? is it some kind of function call?

**Review by fractalcactus — APPROVED** [review]

File: tools/generate-fixture-data.ts [hunk 3]
```diff
@@ -21,8 +22,9 @@ const generateAuditOutput = async (
 ): Promise<string> =>
   new Promise<string>((resolve, reject) => {
     const proc = spawn(
-      packageManager,
+      'npx',
```
Comment (mischa-s) [inline tools/generate-fixture-data.ts:25]: 🎉

**Review by mischa-s — APPROVED** [review]
✨

File: src/types.ts (hunk 2 above)
Comment (mischa-s) [reply src/types.ts:6]: this repo is written in typescript - javascripts strongly typed cousin

Comment (G-Rath) [reply src/types.ts:6]: It's a [generic](https://www.typescriptlang.org/docs/handbook/generics.html) - this is actually a TypeScript thing, which is a superset of Javascript.

Type code like this only "exists" at compile type - the runtime code is regular JS (which this is not, so its removed by the compiler).

Specifically, generics are like parameters for types - they let you combine types together without having to know all the details of what a user is passing in.

A great classic example is arrays, which have type `Array<T>` - You don't actually care what type `T` is, your array methods (`push`, `shift`, `slice`, `pop`, `size`, etc) will work regardless of what the array actually holds.

So the "type" for an Array class looks something like this:

```
class Array<T> {
  push(item: T): void;
  pop(): T;
  includes(item: T): boolean;
  length: number;
}
```

So if you have an array of strings (`Array<string>`), the type for i.e `push` "becomes" `push(item: string): void` :)

Comment (fractalcactus) [reply src/types.ts:6]: I see! thanks everyone

**Pull request merged**